src/libraries/System.Private.CoreLib/src/System/Buffer.cs

```diff
@@ -240,6 +241,39 @@ internal static void Memmove(ref byte dest, ref byte src, nuint len)
             return;
 
         MCPY05:
+            // Misaligned access via SIMD is especially expensive on ARM64 on large data.
+            // 512 is an arbitrary threshold picked for Ampere and Apple M1.
+            //
+            // TODO: Consider doing the same on x86/AMD64 for V256 and V512
+#if HAS_CUSTOM_BLOCKS && TARGET_ARM64
+            if (Vector128.IsHardwareAccelerated && len >= 512)
+            {
+                // Try to opportunistically align the destination below. The input isn't pinned, so the GC
+                // is free to move the references. We're therefore assuming that reads may still be unaligned.
+                //
+                // dest is more important to align than src because an unaligned store is more expensive
+                // than an unaligned load.
+                nuint misalignedElements = (nuint)Unsafe.AsPointer(ref dest) & (Vector128.Size - 1);
+                if (misalignedElements != 0)
+                {
+                    // E.g. if misalignedElements is 4, it means we need to use a scalar loop
+                    // for 16 - 4 = 12 elements till we're aligned to 16b boundary.
+                    misalignedElements = Vector128.Size - misalignedElements;
+                    nuint offset = 0;
+                    do
+                    {
+                        // For large misalignment on x64 we might want to use smaller SIMD here.
+                        Unsafe.Add(ref dest, offset) = Unsafe.Add(ref src, offset);
+                        offset++;
+                    }
+                    while (offset != misalignedElements);
+
+                    src = ref Unsafe.Add(ref src, misalignedElements);
+                    dest = ref Unsafe.Add(ref dest, misalignedElements);
+                    len -= misalignedElements;
+                }
+            }
+#endif
```

Comment:
```suggestion
#if HAS_CUSTOM_BLOCKS
            if (Vector128.IsHardwareAccelerated && len >= 128)
            {
                // Try to opportunistically align the destination below. The input isn't pinned, so the GC
                // is free to move the references. We're therefore assuming that reads may still be unaligned.
                //
                // dest is more important to align than src because an unaligned store is more expensive
                // than an unaligned load.
                nuint misalignedElements = (nuint)Unsafe.AsPointer(ref dest) & (Vector128.Size - 1);
				Vector128.LoadUnsafe(ref src).StoreUnsafe(ref dest);
                src = ref Unsafe.Add(ref src, misalignedElements);
                dest = ref Unsafe.Add(ref dest, misalignedElements);
                len -= misalignedElements;
            }
#endif
```
I feel like there's no sense to be so restrictive here, 128 should probably already have benefit here and from what I've seen XArch benefits from alignment too (especially on AMD CPUs) and there's no reason not to use a single SIMD load to get the alignment here

Reply:
XArch won't benefit from aligning to 16 bytes, only SSE-only, the rest will likely regress.